test/e2e/storage/pvc_protection.go

```diff
@@ -0,0 +1,125 @@
+	"k8s.io/kubernetes/test/e2e/framework"
+)
+
+var _ = SIGDescribe("PVC Protection [Feature:PVCProtection]", func() {
+	var (
+		client                  clientset.Interface
+		nameSpace               string
+		err                     error
+		pvc                     *v1.PersistentVolumeClaim
+		pvcCreatedAndNotDeleted bool
+	)
+
+	f := framework.NewDefaultFramework("pvc-protection")
+	BeforeEach(func() {
+		client = f.ClientSet
+		nameSpace = f.Namespace.Name
+		framework.ExpectNoError(framework.WaitForAllNodesSchedulable(client, framework.TestContext.NodeSchedulableTimeout))
+
+		By("Creating a PVC")
+		suffix := "pvc-protection"
+		defaultSC := getDefaultStorageClassName(client)
+		testStorageClass := storageClassTest{
+			claimSize: "1Gi",
+		}
+		pvc = newClaim(testStorageClass, nameSpace, suffix)
+		pvc.Spec.StorageClassName = &defaultSC
+		pvc, err = client.CoreV1().PersistentVolumeClaims(pvc.Namespace).Create(pvc)
+		Expect(err).NotTo(HaveOccurred(), "Error creating PVC")
+		pvcCreatedAndNotDeleted = true
+
+		By("Waiting for PVC to become Bound")
+		err = framework.WaitForPersistentVolumeClaimPhase(v1.ClaimBound, client, nameSpace, pvc.Name, framework.Poll, framework.ClaimProvisionTimeout)
+		Expect(err).NotTo(HaveOccurred(), "Failed waiting for PVC to be bound %v", err)
+
+		By("Checking that PVC Protection finalizer is set")
+		pvc, err = client.CoreV1().PersistentVolumeClaims(pvc.Namespace).Get(pvc.Name, metav1.GetOptions{})
+		Expect(err).NotTo(HaveOccurred(), "While getting PVC status")
+		Expect(slice.ContainsString(pvc.ObjectMeta.Finalizers, volumeutil.PVCProtectionFinalizer, nil)).To(BeTrue())
+	})
+
+	AfterEach(func() {
+		if pvcCreatedAndNotDeleted {
+			framework.DeletePersistentVolumeClaim(client, pvc.Name, nameSpace)
+		}
+	})
+
+	It("Verify \"immediate\" deletion of a PVC that is not in active use by a pod", func() {
+		By("Deleting the PVC")
+		err = client.CoreV1().PersistentVolumeClaims(pvc.Namespace).Delete(pvc.Name, metav1.NewDeleteOptions(0))
+		Expect(err).NotTo(HaveOccurred(), "Error deleting PVC")
+		waitForPersistentVolumeClaimBeRemoved(client, pvc.Namespace, pvc.Name, framework.Poll, framework.ClaimProvisionTimeout)
+		pvcCreatedAndNotDeleted = false
+	})
+
+	It("Verify that PVC in active use by a pod is not removed immediatelly", func() {
+		By("Creating a Pod that becomes Running and therefore is actively using the PVC")
+		pvcClaims := []*v1.PersistentVolumeClaim{pvc}
+		pod, err := framework.CreatePod(client, nameSpace, nil, pvcClaims, false, "")
+		Expect(err).NotTo(HaveOccurred(), "While creating pod that uses the PVC or waiting for the Pod to become Running")
+
+		By("Deleting the PVC, however, the PVC must not be removed from the system as it's in active use by a pod")
+		err = client.CoreV1().PersistentVolumeClaims(pvc.Namespace).Delete(pvc.Name, metav1.NewDeleteOptions(0))
+		Expect(err).NotTo(HaveOccurred(), "Error deleting PVC")
+
+		By("Checking that the PVC status is Terminating")
```

Comment:
Would it be worth it to wait a little bit to make sure it stays Terminating?

Reply:
@msau42 it's a good question. Actually, I plan to extend this test after [changes into scheduler](https://github.com/kubernetes/kubernetes/pull/55957) are merged. After the extension the test will check that the PVC is Terminating, then it will try to start a pod that uses the Terminating PVC, check that the pod is not scheduled, delete the not-scheduled pod and then verify that the PVC status is still Terminating. So it will make sure that the PVC stays Terminating for some time.